.github/workflows: Refactor CI jobs

thomaseizinger opened the pull request:
## Description

We refactor our continuous integration workflow with the following goals in mind:

- Run as few jobs as possible
- Have the jobs finish as fast as possible
- Have the jobs redo as little work as possible

There are only so many jobs that GitHub Actions will run in parallel.
Thus, it makes sense to not create massive matrices but instead group
things together meaningfully.

The new `test` job will:

- Run once for each crate
- Ensure that the crate compiles on its specified MSRV
- Ensure that the tests pass
- Ensure that there are no semver violations

This is an improvement to before because we are running all of these
in parallel which speeds up execution and highlights more errors at
once. Previously, tests run later in the pipeline would not get run
at all until you make sure the "first" one passes.

We also previously did not verify the MSRV of each crate, making the
setting in the `Cargo.toml` rather pointless.

The new `cross` job supersedes the existing `wasm` job.

This is an improvement because we now also compile the crate for
windows and MacOS. Something that wasn't checked before.
We assume that checking MSRV and the tests under Linux is good enough.
Hence, this job only checks for compile-errors.

The new `feature_matrix` ensures we compile correctly with certain feature combinations.

`libp2p` exposes a fair few feature-flags. Some of the combinations
are worth checking independently. For the moment, this concerns only
the executor related transports together with the executor flags but
this list can easily be extended.

The new `clippy` job runs for `stable` and `beta` rust.

Clippy gets continuously extended with new lints. Up until now, we would only
learn about those as soon as a new version of Rust is released and CI would
run the new lints. This leads to unrelated failures in CI. Running clippy on with `beta`
Rust gives us a heads-up of 6 weeks before these lints land on stable.

Fixes #2951.

## Links to any relevant issues

Depends-On: https://github.com/libp2p/rust-libp2p/pull/3131

## Open Questions

We could merge `cross` and `feature_matrix` together by adding another matrix variable that is either set to `--all-features` or `--features="mdns ..."`. Should we?

<!-- Unresolved questions, if any. -->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

thomaseizinger committed:
Rename CI job

To reuse this, we rename it to what it actually does: Gathering
all crates that are published to crates.io.

thomaseizinger committed:
Refactor test and build jobs

Overall, we want to:

- Run as few jobs as possible
- Have the jobs finish as fast as possible
- Have the jobs redo as little work as possible

There are only so many jobs that GitHub Actions will run in parallel.
Thus, it makes sense to not create massive matrices but instead group
things together meaningfully.

The new `test` job will:

- Run once for each crate
- Ensure that the crate compiles on its specified MSRV
- Ensure that the tests pass
- Ensure that there are no semver violations

This is an improvement to before because we are running all of these
in parallel which speeds up execution and highlights more errors at
once. Previously, tests run later in the pipeline would not get run
at all until you make sure the "first" one passes.

The new `cross` job supersedes the existing `wasm` job.

This is an improvement because we now also compile the crate for
windows and MacOS. Something that wasn't checked before.

`libp2p` exposes a fair few feature-flags. Some of the combinations
are worth checking independently. For the moment, this concerns only
the executor related transports together with the executor flags but
this list can easily be extended.

thomaseizinger committed:
Run clippy on stable and beta

thomaseizinger committed:
Don't install things that are already present

thomaseizinger committed:
Run yaml formatter

thomaseizinger committed:
Set MSRV correctly for all crates

thomaseizinger committed:
Activate necessary features for test

elenaf9 commented on core/Cargo.toml:
```diff
@@ -1,7 +1,7 @@
 name = "libp2p-core"
 edition = "2021"
-rust-version = "1.56.1"
+rust-version = "1.60.0"
```
Why are we changing the rust-version on some packages to `1.60.0` and on others to `1.62.0`?

thomaseizinger replied:
Some only require 1.60, others require 1.62 :)

thomaseizinger committed:
Don't include TLS on wasm platforms

thomaseizinger committed:
Fix clippy warnings

This includes warnings from the latest stable clippy but also for
the next beta.

thomaseizinger committed:
Merge branch 'master' into 2951-ci-matrix

galargh commented on .github/workflows/ci.yml:
```diff
@@ -11,175 +11,229 @@ concurrency:
-    - name: Install Protoc
-      uses: arduino/setup-protoc@v1
+      - name: Install Protoc
+        uses: arduino/setup-protoc@v1
```
If we want to pin to the exact sha, it'd be:
```suggestion
        uses: arduino/setup-protoc@64c0c85d18e984422218383b81c52f8b077404d3 # v1.1.2
```

galargh commented on .github/workflows/ci.yml:
```diff
@@ -11,175 +11,229 @@ concurrency:
+          
+          test "$ALL_FEATURES" = "$FULL_FEATURE"
+
+  gather_published_crates:
```
Since `test` depends on it, my suggestion would be to place this job above it. Not a requirement but it'd make it easier to read.

thomaseizinger replied:
I usually follow a different philosophy:

The most important ~functions~ jobs should be at the top. Assuming sufficiently descriptive names, you shouldn't need the implementation of a ~function~ job to continue reading and understanding what is happening.

If you do the reverse  - putting all "leaf"[^1] jobs at the top - you end up reading a lot of unrelated snippets  before getting to the parts that actually matter.

[^1]: As in, the leafs in a dependency-tree.

thomaseizinger committed:
Pin protoc action

thomaseizinger committed:
Actually adjust name

thomaseizinger committed:
Merge branch 'master' into 2951-ci-matrix

thomaseizinger committed:
Merge branch 'master' into 2951-ci-matrix

thomaseizinger committed:
Ensure we can compile without any features

thomaseizinger committed:
Cache build artefacts for MSRV check

thomaseizinger committed:
Add comment explaining why this caching works

thomaseizinger committed:
Merge branch 'master' into 2951-ci-matrix

thomaseizinger committed:
Remove unnecessary cast (clippy beta lint)

thomaseizinger committed:
Activate default features to make it compile

thomaseizinger committed:
Add crate name to cache key

thomaseizinger committed:
Set missing MSRV for libp2p-quic

jxs commented on .github/workflows/ci.yml:
```diff
@@ -101,6 +101,13 @@ jobs:
+      matrix:
+        rust-version: [
+          stable,
+          beta
```
If we start running `clippy` on `beta` doesn't it make it redundant to run on `stable`? I.e what could be the use case that `stable` finds lints that `beta` doesn't?

thomaseizinger replied:
The `stable` job will be required to be green. The `beta` one can fail!

If we only run beta, we have the same problem where lints appear out of nowhere and fail our CI everytime beta gets cut.

jxs replied:
ah great! Thanks for clarifying Thomas :)

jxs reviewed (approved):
Sorry for the delay Thomas.  
LGTM, superb work :ok_hand:

mxinden reviewed (approved):
Looks good to me. Thanks for putting so much time into making our life easier!

If I understand the changes to the `rust-version` field correctly, they are just updating the field to the reality, i.e. we depended on these versions before but didn't make it explicit. Thus this is not a breaking change.

@thomaseizinger what do you think of mentioning this in the root CHANGELOG.md anyways? Just so that there is no surprise for people.

thomaseizinger committed:
Add changelog entries

thomaseizinger committed:
Merge branch 'master' into 2951-ci-matrix

mxinden committed:
Merge branch 'master' into 2951-ci-matrix

thomaseizinger committed:
Make `libp2p-swarm` wasm compatible again

thomaseizinger committed:
Merge branch 'master' into 2951-ci-matrix

thomaseizinger committed:
Set MSRV for webrtc

thomaseizinger committed:
Fix clippy beta lints

thomaseizinger committed:
Don't compile libp2p-webrtc for wasm

mxinden committed:
Merge branch 'master' into 2951-ci-matrix
